File: Golang/637. Average of Levels in Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func averageOfLevels(root *TreeNode) []float64 {
    layer := 0
    nums := []float64{}
    counts := []float64{}
    dealNode(root, layer, &nums, &counts)
    for i, _ := range nums {
        nums[i] /= counts[i]
    }
    return nums
}

func dealNode(node *TreeNode, layer int, nums, counts *[]float64) {
    if node != nil {
        if layer < len(*nums) {
            (*nums)[layer] += float64(node.Val)
            (*counts)[layer]++ 
        } else {
            *nums = append(*nums, float64(node.Val))
            *counts = append(*counts, 1)
        }
        dealNode(node.Left, layer+1, nums, counts)
        dealNode(node.Right, layer+1, nums, counts)
    }
}

File: Python3/637. Average of Levels in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        nums = []
        counts = []
        self.dealNode(root, 0, nums, counts)
        return [x/float(y) for (x, y) in zip(nums, counts)]
    
    def dealNode(self, node, layer, nums, counts):
        if node:
            if layer < len(nums):
                nums[layer] += node.val
                counts[layer] += 1
            else:
                nums.append(node.val)
                counts.append(1)
            self.dealNode(node.left, layer+1, nums, counts)
            self.dealNode(node.right, layer+1, nums, counts)
